src/agent.c

```diff
@@ -0,0 +1,1719 @@
+                ret = wc_Sha256Update(&sha, data, dataSz);
+            if (ret == 0) {
+                ret = wc_Sha256Final(&sha, digest);
+                ret = (ret == 0) ? WS_SUCCESS : WS_CRYPTO_FAILED;
+            }
+        }
+        else
+            ret = WS_INVALID_ALGO_ID;
+    }
+
+    if (ret == WS_SUCCESS) {
+        byte sig[256];
+        int sigSz = sizeof(sig);
+
+        if (cur->keyType == ID_SSH_RSA) {
+            WOLFSSH_AGENT_KEY_RSA* key;
+            RsaKey rsa;
+            byte encSig[MAX_ENCODED_SIG_SZ];
+            int encSigSz;
+            enum wc_HashType hashType = WC_HASH_TYPE_NONE;
+
+            key = &cur->key.rsa;
+
+            if (flags & AGENT_SIGN_RSA_SHA2_256)
+                hashType = WC_HASH_TYPE_SHA256;
+            else if (flags & AGENT_SIGN_RSA_SHA2_512)
+                hashType = WC_HASH_TYPE_SHA512;
+
+            wc_InitRsaKey(&rsa, agent->heap);
+            mp_read_unsigned_bin(&rsa.n, key->n, key->nSz);
+            mp_read_unsigned_bin(&rsa.e, key->e, key->eSz);
+            mp_read_unsigned_bin(&rsa.d, key->d, key->dSz);
+            mp_read_unsigned_bin(&rsa.p, key->p, key->pSz);
+            mp_read_unsigned_bin(&rsa.q, key->q, key->qSz);
+            mp_read_unsigned_bin(&rsa.u, key->iqmp, key->iqmpSz);
+
+            encSigSz = wc_EncodeSignature(encSig, digest,
+                                          wc_HashGetDigestSize(hashType),
+                                          wc_HashGetOID(hashType));
+            if (encSigSz <= 0) {
+                WLOG(WS_LOG_DEBUG, "Bad Encode Sig");
+                ret = WS_CRYPTO_FAILED;
+            }
+            else {
+                WLOG(WS_LOG_INFO, "Signing hash with RSA.");
+                sigSz = wc_RsaSSL_Sign(encSig, encSigSz, sig, sizeof(sig),
+                                       &rsa, &agent->rng);
+                if (sigSz <= 0) {
+                    WLOG(WS_LOG_DEBUG, "Bad RSA Sign");
+                    ret = WS_RSA_E;
+                }
+            }
+
+            wc_FreeRsaKey(&rsa);
+            if (ret != 0)
+                ret = WS_RSA_E;
+        }
+        else if (cur->keyType == ID_ECDSA_SHA2_NISTP256) {
+            WOLFSSH_AGENT_KEY_ECDSA* key;
+            ecc_key ecc;
+            enum wc_HashType hashType = WC_HASH_TYPE_SHA256;
+
+            key = &cur->key.ecdsa;
+            ret = wc_ecc_import_private_key_ex(key->d, key->dSz,
+                    key->q, key->qSz,
+                    &ecc, ECC_SECP256R1);
+
+            if (ret == 0) {
+                ret = wc_ecc_sign_hash(digest, wc_HashGetDigestSize(hashType),
+                        sig, (word32*)&sigSz, &agent->rng, &ecc);
+            }
+            if (ret != 0)  {
+                WLOG(WS_LOG_DEBUG, "SendKexDhReply: Bad ECDSA Sign");
+                ret = WS_ECC_E;
+            }
+            else {
+                byte r[257];
+                word32 rSz = sizeof(r);
+                byte rPad;
+                byte s[257];
```

Comment:
257 seems large for r and s? can we use MAX_ECC_BYTES here?

Reply:
I'm going to use MAX_ECC_BYTES + ECC_MAX_PAD_SZ, which will be 258.